### PR TITLE
Ro volatile byte kcontrol

### PR DESCRIPTION
**Pull request by RDharageswari**
This patch set allows a user to create Readonly and writeonly kcontrols. It also adds the snd_sof_bytes_ext_get_pm IO control to get the "actual values" of the requested parameters from DSP. These patches mainly help for the tuning purpose.

### CODE REVIEW

File: sound/soc/soc-topology.c [hunk 1]
```diff
@@ -578,7 +578,9 @@ static int soc_tplg_kcontrol_bind_io(struct snd_soc_tplg_ctl_hdr *hdr,
 
 	if (le32_to_cpu(hdr->ops.info) == SND_SOC_TPLG_CTL_BYTES
 		&& k->iface & SNDRV_CTL_ELEM_IFACE_MIXER
-		&& k->access & SNDRV_CTL_ELEM_ACCESS_TLV_READWRITE
+		&& (k->access & SNDRV_CTL_ELEM_ACCESS_TLV_READWRITE
+		|| k->access & SNDRV_CTL_ELEM_ACCESS_TLV_WRITE
+		|| k->access & SNDRV_CTL_ELEM_ACCESS_TLV_READ)
```
Comment (bardliao) [inline sound/soc/soc-topology.c]: ``k->access & SNDRV_CTL_ELEM_ACCESS_TLV_READWRITE`` should be enough since ``#define SNDRV_CTL_ELEM_ACCESS_TLV_READWRITE     (SNDRV_CTL_ELEM_ACCESS_TLV_READ|SNDRV_CTL_ELEM_ACCESS_TLV_WRITE)``

File: sound/soc/sof/sof-audio.h [hunk 2]
```diff
@@ -139,6 +139,9 @@ int snd_sof_bytes_ext_put(struct snd_kcontrol *kcontrol,
 int snd_sof_bytes_ext_get(struct snd_kcontrol *kcontrol,
 			  unsigned int __user *binary_data,
 			  unsigned int size);
+int snd_sof_bytes_ext_volatile_get(struct snd_kcontrol *kcontrol,
+			  unsigned int __user *binary_data,
```
Comment (bardliao) [inline sound/soc/sof/sof-audio.h]: Alignment should match open parenthesis, please.

File: sound/soc/sof/control.c [hunk 3]
```diff
@@ -353,6 +353,78 @@ int snd_sof_bytes_ext_put(struct snd_kcontrol *kcontrol,
...
+	struct sof_ipc_ctrl_data *cdata = scontrol->control_data;
+	struct snd_ctl_tlv header;
+	struct snd_ctl_tlv __user *tlvd =
+		(struct snd_ctl_tlv __user *)binary_data;
+	int data_size;
+	int ret = 0;
+	int err;
+
+	ret = pm_runtime_get_sync(scomp->dev);
+        if (ret < 0) {
```
Comment (bardliao) [inline sound/soc/sof/control.c]: code indent should use tabs where possible, please

File: sound/soc/soc-topology.c [hunk 4]
```diff
@@ -603,7 +605,14 @@ static int soc_tplg_kcontrol_bind_io(struct snd_soc_tplg_ctl_hdr *hdr,
 				sbe->get = ext_ops[i].get;
 		}
 
-		if (sbe->put && sbe->get)
+		if ((k->access & SNDRV_CTL_ELEM_ACCESS_TLV_READWRITE)
```
Comment (ranj063) [inline sound/soc/soc-topology.c]: this will be true for both READ_ONLY and WRITE_ONLY accesses as well. So it should be changed to 
(k->access & SNDRV_CTL_ELEM_ACCESS_TLV_READWRITE == SNDRV_CTL_ELEM_ACCESS_TLV_READWRITE)

File: sound/soc/sof/control.c [hunk 5]
```diff
@@ -353,6 +353,78 @@ int snd_sof_bytes_ext_put(struct snd_kcontrol *kcontrol,
...
+	/*
+	 * Decrement the limit by ext bytes header size to
+	 * ensure the user space buffer is not exceeded.
+	 */
+	size -= sizeof(const struct snd_ctl_tlv);
+
+	/* set the ABI header values */
+	cdata->data->magic = SOF_ABI_MAGIC;
+	cdata->data->abi = SOF_ABI_VERSION;
+	        /* get all the component data from DSP */
```
Comment (ranj063) [inline sound/soc/sof/control.c]: alignment?

File: sound/soc/sof/control.c [hunk 6]
```diff
@@ -353,6 +353,78 @@ int snd_sof_bytes_ext_put(struct snd_kcontrol *kcontrol,
...
+			  unsigned int size)
+{
+	struct soc_bytes_ext *be =
+		(struct soc_bytes_ext *)kcontrol->private_value;
+	struct snd_sof_control *scontrol = be->dobj.private;
+	struct snd_soc_component *scomp = scontrol->scomp;
+	struct sof_ipc_ctrl_data *cdata = scontrol->control_data;
+	struct snd_ctl_tlv header;
+	struct snd_ctl_tlv __user *tlvd =
+		(struct snd_ctl_tlv __user *)binary_data;
```
Comment (ranj063) [inline sound/soc/sof/control.c]: this should  fit on the previous line?

File: sound/soc/sof/control.c [hunk 7]
```diff
@@ -353,6 +353,78 @@ int snd_sof_bytes_ext_put(struct snd_kcontrol *kcontrol,
...
+	int data_size;
+	int ret = 0;
+	int err;
+
+	ret = pm_runtime_get_sync(scomp->dev);
+        if (ret < 0) {
+                dev_err_ratelimited(scomp->dev,
+                                    "error: bytes_ext get failed to resume %d\n",
+                                    ret);
```
Comment (ranj063) [inline sound/soc/sof/control.c]: same here. should all fit in one line (use all 100 chars)

File: sound/soc/sof/control.c [hunk 8]
```diff
@@ -353,6 +353,78 @@ int snd_sof_bytes_ext_put(struct snd_kcontrol *kcontrol,
...
+
+	/* set the ABI header values */
+	cdata->data->magic = SOF_ABI_MAGIC;
+	cdata->data->abi = SOF_ABI_VERSION;
+	        /* get all the component data from DSP */
+        ret = snd_sof_ipc_set_get_comp_data(scontrol,
+                                            SOF_IPC_COMP_GET_DATA,
+                                            SOF_CTRL_TYPE_DATA_GET,
+                                            scontrol->cmd,
+                                            false);
```
Comment (ranj063) [inline sound/soc/sof/control.c]: can you please condense this also to use 100 chars per line too?

File: sound/soc/sof/control.c [hunk 9]
```diff
@@ -353,6 +353,78 @@ int snd_sof_bytes_ext_put(struct snd_kcontrol *kcontrol,
...
+                                    ret);
+                pm_runtime_put_noidle(scomp->dev);
+                return ret;
+        }
+
+	/*
+	 * Decrement the limit by ext bytes header size to
+	 * ensure the user space buffer is not exceeded.
+	 */
+	size -= sizeof(const struct snd_ctl_tlv);
```
Comment (ranj063) [inline sound/soc/sof/control.c]: why is this needed? It doesnt seem to be used anywhere?

File: sound/soc/sof/control.c [hunk 10]
```diff
@@ -353,6 +353,65 @@ int snd_sof_bytes_ext_put(struct snd_kcontrol *kcontrol,
 	return 0;
 }
 
+
+int snd_sof_bytes_ext_volatile_get(struct snd_kcontrol *kcontrol,
```
Comment (ranj063) [inline sound/soc/sof/control.c]: the commit message has a different name for the function.

File: sound/soc/sof/control.c [hunk 11]
```diff
@@ -353,6 +353,65 @@ int snd_sof_bytes_ext_put(struct snd_kcontrol *kcontrol,
...
+			  unsigned int __user *binary_data,
+			  unsigned int size)
+{
+	struct soc_bytes_ext *be =
+		(struct soc_bytes_ext *)kcontrol->private_value;
```
Comment (ranj063) [inline sound/soc/sof/control.c]: doesnt fit in one line?

File: sound/soc/sof/control.c [hunk 12]
```diff
@@ -353,6 +353,65 @@ int snd_sof_bytes_ext_put(struct snd_kcontrol *kcontrol,
...
+
+	/* set the ABI header values */
+	cdata->data->magic = SOF_ABI_MAGIC;
+	cdata->data->abi = SOF_ABI_VERSION;
+	/* get all the component data from DSP */
+	ret = snd_sof_ipc_set_get_comp_data(scontrol, SOF_IPC_COMP_GET_DATA,
+					    SOF_CTRL_TYPE_DATA_GET, scontrol->cmd, false);
+
+	/* Prevent read of other kernel data or possibly corrupt response */
+	data_size = cdata->data->size + sizeof(const struct sof_abi_hdr);
```
Comment (ranj063) [inline sound/soc/sof/control.c]: should you also include the size of tlv header in this calculation?

File: sound/soc/soc-topology.c [hunk 13]
```diff
@@ -603,10 +603,16 @@ static int soc_tplg_kcontrol_bind_io(struct snd_soc_tplg_ctl_hdr *hdr,
...
+		     == SNDRV_CTL_ELEM_ACCESS_TLV_READWRITE) && (sbe->put && sbe->get))
 			return 0;
-		else
-			return -EINVAL;
+		if ((k->access & SNDRV_CTL_ELEM_ACCESS_TLV_READ) && sbe->get)
+			return 0;
+		if ((k->access & SNDRV_CTL_ELEM_ACCESS_TLV_WRITE) && sbe->put)
+			return 0;
+
+		return -EINVAL;
```
Comment (kv2019i) [inline sound/soc/soc-topology.c]: @RDharageswari Couple of questions. On line 582 there's a check for TLV_READWRITE, so can this code be reached with only TLV_READ set? Maybe that needs to be changed as well. To make the code a bit simpler, maybe you can reverse the logic and check for:
```
if (TLV_READ && !sbe->get)
        return -EINVAL;
if (TLV_WRITE && !sbe->put)
       return -EINVAL;
```

TLV_READWRITE is defined as bit OR of READ and WRITE bits, so above should cover also the original check for both put and get.

File: sound/soc/soc-topology.c [hunk 14]
```diff
@@ -623,7 +629,13 @@ static int soc_tplg_kcontrol_bind_io(struct snd_soc_tplg_ctl_hdr *hdr,
...
 
 	/* vendor specific handlers found ? */
-	if (k->put && k->get && k->info)
+	/* RW controls need to have both get/put IO handlers */
+	if (((k->access & SNDRV_CTL_ELEM_ACCESS_TLV_READWRITE)
+	     == SNDRV_CTL_ELEM_ACCESS_TLV_READWRITE) && (k->put && k->get))
+		return 0;
+	if ((k->access & SNDRV_CTL_ELEM_ACCESS_TLV_READ) && k->get)
+		return 0;
+	if ((k->access & SNDRV_CTL_ELEM_ACCESS_TLV_WRITE) && k->put)
```
Comment (kv2019i) [inline sound/soc/soc-topology.c]: The check for "k->info" is dropped here. Maybe use a simpler logic like above and then it's easier to check for "k->info" as well.

File: sound/soc/soc-topology.c [hunk 15]
```diff
@@ -640,7 +652,13 @@ static int soc_tplg_kcontrol_bind_io(struct snd_soc_tplg_ctl_hdr *hdr,
...
 
 	/* standard handlers found ? */
-	if (k->put && k->get && k->info)
+	/* RW controls need to have both get/put IO handlers */
+	if (((k->access & SNDRV_CTL_ELEM_ACCESS_TLV_READWRITE)
+	     == SNDRV_CTL_ELEM_ACCESS_TLV_READWRITE) && (k->put && k->get))
+		return 0;
+	if ((k->access & SNDRV_CTL_ELEM_ACCESS_TLV_READ) && k->get)
+		return 0;
+	if ((k->access & SNDRV_CTL_ELEM_ACCESS_TLV_WRITE) && k->put)
```
Comment (kv2019i) [inline sound/soc/soc-topology.c]: "k->info" check missing here as well.

File: sound/soc/sof/control.c [hunk 16]
```diff
@@ -353,6 +353,65 @@ int snd_sof_bytes_ext_put(struct snd_kcontrol *kcontrol,
...
+		dev_err_ratelimited(scomp->dev, "error: bytes_ext get failed to resume %d\n", ret);
+		pm_runtime_put_noidle(scomp->dev);
+		return ret;
+	}
+
+	/* set the ABI header values */
+	cdata->data->magic = SOF_ABI_MAGIC;
+	cdata->data->abi = SOF_ABI_VERSION;
+	/* get all the component data from DSP */
+	ret = snd_sof_ipc_set_get_comp_data(scontrol, SOF_IPC_COMP_GET_DATA,
```
Comment (kv2019i) [inline sound/soc/sof/control.c]: @RDharageswari Better to check for errors here and not write out garbage to user-space on IPC error.

File: sound/soc/sof/control.c (hunk 12 above)
Comment (RDharageswari) [reply sound/soc/sof/control.c]: Hi Ranjani,
I dont think TLV header is needed here. For get params IPC, fw sends the response as sof_abi_hdr + model data.
cdata->data->size is just the model data size. So, for the condition check here, cdata->data->size + sizeof(const struct sof_abi_hdr) becomes mandatory(data_size > be->max) be->max is from topology the max control size.

data_size will just hold the actual data size from the firmware(ABI_header + model_data).
The application reading data while allocating the buffer to read will add the size for tlv header. 
eg: sof-ctl:
/*
         * Allocate buffer for tlv write/read. The buffer needs a two
         * words header with tag (SOF_CTRL_CMD_BINARY) and size in bytes.
         */
        buffer_size = ctl_data->ctrl_size + 2 * sizeof(unsigned int);

File: sound/soc/soc-topology.c (hunk 13 above)
Comment (RDharageswari) [reply sound/soc/soc-topology.c]: if i understand your question correctly, yes this piece of code is valid for RW, RO, WO any of the scenarios

File: sound/soc/soc-topology.c (hunk 15 above)
Comment (RDharageswari) [reply sound/soc/soc-topology.c]: Thanks, added

File: sound/soc/sof/control.c (hunk 16 above)
Comment (RDharageswari) [reply sound/soc/sof/control.c]: Thanks added

File: sound/soc/sof/control.c (hunk 12 above)
Comment (ranj063) [reply sound/soc/sof/control.c]: thanks for the explanation @RDharageswari .In that case the check looks fine.

File: sound/soc/soc-topology.c [hunk 17]
```diff
@@ -623,7 +623,13 @@ static int soc_tplg_kcontrol_bind_io(struct snd_soc_tplg_ctl_hdr *hdr,
 	}
 
 	/* vendor specific handlers found ? */
-	if (k->put && k->get && k->info)
+	/* RW controls need to have both get/put IO handlers */
```
Comment (ranj063) [inline sound/soc/soc-topology.c]: @RDharageswari these checks are for the io_ops and not for the ext_ops right? Do we really need to check these here?

File: sound/soc/sof/control.c [hunk 18]
```diff
@@ -353,6 +353,66 @@ int snd_sof_bytes_ext_put(struct snd_kcontrol *kcontrol,
...
+			  unsigned int size)
+{
+	struct soc_bytes_ext *be = (struct soc_bytes_ext *)kcontrol->private_value;
+	struct snd_sof_control *scontrol = be->dobj.private;
+	struct snd_soc_component *scomp = scontrol->scomp;
+	struct sof_ipc_ctrl_data *cdata = scontrol->control_data;
+	struct snd_ctl_tlv header;
+	struct snd_ctl_tlv __user *tlvd = (struct snd_ctl_tlv __user *)binary_data;
+	int data_size;
+	int ret = 0;
```
Comment (ranj063) [inline sound/soc/sof/control.c]: no need for initializing yet

File: include/uapi/sound/sof/tokens.h [hunk 19]
```diff
@@ -24,6 +24,9 @@
 #define SOF_TPLG_KCTL_ENUM_ID	257
 #define SOF_TPLG_KCTL_BYTES_ID	258
 #define SOF_TPLG_KCTL_SWITCH_ID	259
+#define SOF_TPLG_KCTL_BYTES_VOLATILE_RO 260
```
Comment (ranj063) [inline include/uapi/sound/sof/tokens.h:27]: @RDharageswari the last sentence in the commit message is not necessary. How it is used depends on the user really

File: sound/soc/sof/control.c (hunk 18 above)
Comment (RDharageswari) [reply sound/soc/sof/control.c]: Done

File: include/uapi/sound/sof/tokens.h (hunk 19 above)
Comment (RDharageswari) [reply include/uapi/sound/sof/tokens.h:27]: done

File: sound/soc/soc-topology.c [hunk 20]
```diff
@@ -591,7 +591,6 @@ static int soc_tplg_kcontrol_bind_io(struct snd_soc_tplg_ctl_hdr *hdr,
 		 */
 		k->info = snd_soc_bytes_info_ext;
 		k->tlv.c = snd_soc_bytes_tlv_callback;
-
```
Comment (ranj063) [inline sound/soc/soc-topology.c]: spurious change.

File: sound/soc/soc-topology.c (hunk 17 above)
Comment (RDharageswari) [reply sound/soc/soc-topology.c]: Removed the handling here

File: sound/soc/soc-topology.c (hunk 20 above)
Comment (RDharageswari) [reply sound/soc/soc-topology.c]: Removed this change

File: sound/soc/soc-topology.c [hunk 21]
```diff
@@ -603,10 +603,11 @@ static int soc_tplg_kcontrol_bind_io(struct snd_soc_tplg_ctl_hdr *hdr,
 				sbe->get = ext_ops[i].get;
 		}
 
-		if (sbe->put && sbe->get)
-			return 0;
-		else
+		if ((k->access & SNDRV_CTL_ELEM_ACCESS_TLV_READ) && !sbe->get)
```
Comment (ranj063) [inline sound/soc/soc-topology.c:606]: @RDharageswari the patch looks good now but can you please simplify the commit message to say something along these lines. There's no need to mention anything about volatile here as you're not really checking that here.

"Add support for RO/WO TLV byte controls

Add support for read-only and write-only byte controls by checking for the appropriate get/put IO handlers"

File: sound/soc/sof/sof-audio.h [hunk 22]
```diff
@@ -139,6 +139,9 @@ int snd_sof_bytes_ext_put(struct snd_kcontrol *kcontrol,
 int snd_sof_bytes_ext_get(struct snd_kcontrol *kcontrol,
 			  unsigned int __user *binary_data,
 			  unsigned int size);
+int snd_sof_bytes_ext_volatile_get(struct snd_kcontrol *kcontrol,
+				   unsigned int __user *binary_data,
+				   unsigned int size);
```
Comment (ranj063) [inline sound/soc/sof/sof-audio.h:143]: please use all 100chars and condense to 2 lines?

File: sound/soc/sof/control.c [hunk 23]
```diff
@@ -353,6 +353,66 @@ int snd_sof_bytes_ext_put(struct snd_kcontrol *kcontrol,
 	return 0;
 }
 
+
+int snd_sof_bytes_ext_volatile_get(struct snd_kcontrol *kcontrol,
+			  unsigned int __user *binary_data,
```
Comment (ranj063) [inline sound/soc/sof/control.c]: same here. use 100 chars and keep to 2 lines

File: sound/soc/soc-topology.c (hunk 21 above)
Comment (RDharageswari) [reply sound/soc/soc-topology.c:606]: Done

File: sound/soc/sof/sof-audio.h (hunk 22 above)
Comment (RDharageswari) [reply sound/soc/sof/sof-audio.h:143]: Done

File: sound/soc/sof/control.c (hunk 23 above)
Comment (RDharageswari) [reply sound/soc/sof/control.c]: Done

**Comment by ranj063** [comment]
@bardliao you good with the changes now?

**Review by bardliao — APPROVED** [review]

File: sound/soc/sof/control.c (hunk 12 above)
Comment (lyakh) [reply sound/soc/sof/control.c]: don't know how paranoid we want to be here, but in principle you have an overflow possibility here. If `cdata->data->size` is returned by the FW as a small negative number (-1), it'll pass the test, but obviously that's invalid.

File: sound/soc/sof/control.c [hunk 24]
```diff
@@ -353,6 +353,65 @@ int snd_sof_bytes_ext_put(struct snd_kcontrol *kcontrol,
...
+int snd_sof_bytes_ext_volatile_get(struct snd_kcontrol *kcontrol, unsigned int __user *binary_data,
+				   unsigned int size)
+{
+	struct soc_bytes_ext *be = (struct soc_bytes_ext *)kcontrol->private_value;
+	struct snd_sof_control *scontrol = be->dobj.private;
+	struct snd_soc_component *scomp = scontrol->scomp;
+	struct sof_ipc_ctrl_data *cdata = scontrol->control_data;
+	struct snd_ctl_tlv header;
+	struct snd_ctl_tlv __user *tlvd = (struct snd_ctl_tlv __user *)binary_data;
+	int data_size;
```
Comment (lyakh) [inline sound/soc/sof/control.c]: `size_t`

**Review by lyakh — CHANGES_REQUESTED** [review]

File: sound/soc/sof/control.c (hunk 24 above)
Comment (RDharageswari) [reply sound/soc/sof/control.c]: Done

**Review by bardliao — APPROVED** [review]

File: sound/soc/sof/control.c [hunk 25]
```diff
@@ -353,6 +353,65 @@ int snd_sof_bytes_ext_put(struct snd_kcontrol *kcontrol,
...
+	ret = snd_sof_ipc_set_get_comp_data(scontrol, SOF_IPC_COMP_GET_DATA, SOF_CTRL_TYPE_DATA_GET,
+					    scontrol->cmd, false);
+	if (ret < 0)
+		goto out;
+
+	/* Prevent read of other kernel data or possibly corrupt response */
+	data_size = cdata->data->size + sizeof(const struct sof_abi_hdr);
+
+	/* check data size doesn't exceed max coming from topology */
+	if (data_size > be->max) {
```
Comment (lyakh) [inline sound/soc/sof/control.c]: the point I was trying to make above is, that this test is incomplete. You're writing, that you want to check for corrupted data, but if, as I suggested, `cdata->data->size == (uint32_t)-1`, which is invalid, you won't catch it. Please have a look at #2383

Comment (RDharageswari) [reply sound/soc/sof/control.c]: Sure

**Review by lyakh — APPROVED** [review]
